--- a/src/google/protobuf/compiler/js/js_generator.h	(index)
+++ b/src/google/protobuf/compiler/js/js_generator.h	(working tree)
@@ -154,7 +154,7 @@ class PROTOC_EXPORT Generator : public CodeGenerator {
                    const TProtoStringType& parameter, GeneratorContext* context,
                    TProtoStringType* error) const override;
 
-  uint64 GetSupportedFeatures() const override {
+  uint64_t GetSupportedFeatures() const override {
     return FEATURE_PROTO3_OPTIONAL;
   }
 
