--- a/src/google/protobuf/stubs/stl_util.h	(index)
+++ b/src/google/protobuf/stubs/stl_util.h	(working tree)
@@ -47,7 +47,7 @@ namespace protobuf {
 // place in open source code.  Feel free to fill this function in with your
 // own disgusting hack if you want the perf boost.
 inline void STLStringResizeUninitialized(TProtoStringType* s, size_t new_size) {
-  s->resize(new_size);
+  s->ReserveAndResize(new_size);
 }

 // As above, but we make sure to follow amortized growth in which we always
