diff --color -bur a/src/google/protobuf/compiler/cpp/cpp_file.cc b/src/google/protobuf/compiler/cpp/cpp_file.cc
--- a/src/google/protobuf/compiler/cpp/cpp_file.cc	2024-02-12 19:22:03.524306801 +0000
+++ b/src/google/protobuf/compiler/cpp/cpp_file.cc	2024-02-12 19:31:19.776317755 +0000
@@ -317,6 +317,17 @@
   GenerateBottomHeaderGuard(printer, true);
 }
 
+void FileGenerator::GeneratePBDeps(io::Printer* printer,
+                                   const TProtoStringType& info_path) {
+  Formatter format(printer, variables_);
+
+  GenerateTopHeaderGuard(printer, true, true);
+
+  GenerateDependencyIncludes(printer);
+
+  GenerateBottomHeaderGuard(printer, true, true);
+}
+
 void FileGenerator::DoIncludeFile(const TProtoStringType& google3_name,
                                   bool do_export, io::Printer* printer) {
   Formatter format(printer, variables_);
@@ -1142,7 +1153,7 @@
   format("PROTOBUF_NAMESPACE_CLOSE\n");
 }
 
-void FileGenerator::GenerateTopHeaderGuard(io::Printer* printer, bool pb_h) {
+void FileGenerator::GenerateTopHeaderGuard(io::Printer* printer, bool pb_h, bool deps) {
   Formatter format(printer, variables_);
   // Generate top of header.
   format(
@@ -1154,7 +1165,7 @@
       "\n"
       "#include <limits>\n"
       "#include <string>\n",
-      IncludeGuard(file_, pb_h, options_));
+      IncludeGuard(file_, pb_h, deps, options_));
   if (!options_.opensource_runtime && !enum_generators_.empty()) {
     // Add header to provide std::is_integral for safe Enum_Name() function.
     format("#include <type_traits>\n");
@@ -1162,10 +1173,10 @@
   format("\n");
 }
 
-void FileGenerator::GenerateBottomHeaderGuard(io::Printer* printer, bool pb_h) {
+void FileGenerator::GenerateBottomHeaderGuard(io::Printer* printer, bool pb_h, bool deps) {
   Formatter format(printer, variables_);
   format("#endif  // $GOOGLE_PROTOBUF$_INCLUDED_$1$\n",
-         IncludeGuard(file_, pb_h, options_));
+         IncludeGuard(file_, pb_h, deps, options_));
 }
 
 void FileGenerator::GenerateLibraryIncludes(io::Printer* printer) {
diff --color -bur a/src/google/protobuf/compiler/cpp/cpp_file.h b/src/google/protobuf/compiler/cpp/cpp_file.h
--- a/src/google/protobuf/compiler/cpp/cpp_file.h	2024-02-12 19:22:03.524306801 +0000
+++ b/src/google/protobuf/compiler/cpp/cpp_file.h	2024-02-12 19:32:03.213555311 +0000
@@ -80,6 +80,7 @@
   // info_path, if non-empty, should be the path (relative to printer's
   // output) to the metadata file describing this PB header.
   void GeneratePBHeader(io::Printer* printer, const TProtoStringType& info_path);
+  void GeneratePBDeps(io::Printer* printer, const TProtoStringType& info_path);
   void GenerateSource(io::Printer* printer);
 
   // The following member functions are used when the lite_implicit_weak_fields
@@ -133,8 +134,8 @@
   void GenerateForwardDeclarations(io::Printer* printer);
 
   // Generates top or bottom of a header file.
-  void GenerateTopHeaderGuard(io::Printer* printer, bool pb_h);
-  void GenerateBottomHeaderGuard(io::Printer* printer, bool pb_h);
+  void GenerateTopHeaderGuard(io::Printer* printer, bool pb_h, bool deps = false);
+  void GenerateBottomHeaderGuard(io::Printer* printer, bool pb_h, bool deps = false);
 
   // Generates #include directives.
   void GenerateLibraryIncludes(io::Printer* printer);
diff --color -bur a/src/google/protobuf/compiler/cpp/cpp_generator.cc b/src/google/protobuf/compiler/cpp/cpp_generator.cc
--- a/src/google/protobuf/compiler/cpp/cpp_generator.cc	2024-02-12 19:22:03.525306806 +0000
+++ b/src/google/protobuf/compiler/cpp/cpp_generator.cc	2024-02-12 19:32:26.781701435 +0000
@@ -214,6 +214,25 @@
     }
   }
 
+  if (!file_options.transitive_pb_h) {
+    std::unique_ptr<io::ZeroCopyOutputStream> output(
+        generator_context->Open(basename + ".deps.pb.h"));
+    GeneratedCodeInfo annotations;
+    io::AnnotationProtoCollector<GeneratedCodeInfo> annotation_collector(
+        &annotations);
+    TProtoStringType info_path = basename + ".deps.pb.h.meta";
+    io::Printer printer(
+        output.get(), '$',
+        file_options.annotate_headers ? &annotation_collector : NULL);
+    file_generator.GeneratePBDeps(
+        &printer, file_options.annotate_headers ? info_path : "");
+    if (file_options.annotate_headers) {
+      std::unique_ptr<io::ZeroCopyOutputStream> info_output(
+          generator_context->Open(info_path));
+      annotations.SerializeToZeroCopyStream(info_output.get());
+    }
+  }
+
   // Generate cc file(s).
   if (UsingImplicitWeakFields(file, file_options)) {
     {
diff --color -bur a/src/google/protobuf/compiler/cpp/cpp_helpers.h b/src/google/protobuf/compiler/cpp/cpp_helpers.h
--- a/src/google/protobuf/compiler/cpp/cpp_helpers.h	2024-02-12 19:22:03.524306801 +0000
+++ b/src/google/protobuf/compiler/cpp/cpp_helpers.h	2024-02-12 19:23:43.999844810 +0000
@@ -478,11 +478,12 @@
 bool IsWellKnownMessage(const FileDescriptor* descriptor);
 
 inline TProtoStringType IncludeGuard(const FileDescriptor* file, bool pb_h,
+                                bool deps,
                                 const Options& options) {
   // If we are generating a .pb.h file and the proto_h option is enabled, then
   // the .pb.h gets an extra suffix.
   TProtoStringType filename_identifier = FilenameIdentifier(
-      file->name() + (pb_h && options.proto_h ? ".pb.h" : ""));
+      file->name() + (deps ? ".deps": "") + (pb_h && options.proto_h ? ".pb.h" : ""));
 
   if (IsWellKnownMessage(file)) {
     // For well-known messages we need third_party/protobuf and net/proto2 to
