--- a/src/google/protobuf/stubs/int128.cc	(index)
+++ b/src/google/protobuf/stubs/int128.cc	(working tree)
@@ -180,7 +180,7 @@ std::ostream& operator<<(std::ostream& o, const uint128& b) {
     os << std::noshowbase << std::setfill('0') << std::setw(div_base_log);
   }
   os << low.lo_;
-  TProtoStringType rep = os.str();
+  std::string rep = os.str();
 
   // Add the requisite padding.
   std::streamsize width = o.width(0);
