--- contrib/libs/virtiofsd/fuse_lowlevel.c	(index)
+++ contrib/libs/virtiofsd/fuse_lowlevel.c	(working tree)
@@ -734,7 +734,7 @@ static void do_forget(fuse_req_t req, fuse_ino_t nodeid,
 
     arg = fuse_mbuf_iter_advance(iter, sizeof(*arg));
     if (!arg) {
-        fuse_reply_err(req, EINVAL);
+        fuse_reply_none(req);
         return;
     }
 
@@ -2506,7 +2506,12 @@ void fuse_session_process_buf_int(struct fuse_session *se,
     return;
 
 reply_err:
-    fuse_reply_err(req, err);
+    if (in->opcode == FUSE_FORGET || in->opcode == FUSE_BATCH_FORGET) {
+        fuse_reply_none(req);
+    } else {
+        fuse_reply_err(req, err);
+    }
+
     pthread_rwlock_unlock(&se->init_rwlock);
 }
 
