diff --git a/include/__algorithm/sort.h b/include/__algorithm/sort.h
index 567c988..258c6a5 100644
--- a/include/__algorithm/sort.h
+++ b/include/__algorithm/sort.h
@@ -954,18 +954,6 @@ void __sort_impl(_RandomAccessIterator __first, _RandomAccessIterator __last, _C
   std::__check_strict_weak_ordering_sorted(std::__unwrap_iter(__first), std::__unwrap_iter(__last), __comp);
 }
 
-template <class _RandomAccessIterator, class _Comp>
-inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20
-void sort(_RandomAccessIterator __first, _RandomAccessIterator __last, _Comp __comp) {
-  std::__sort_impl<_ClassicAlgPolicy>(std::move(__first), std::move(__last), __comp);
-}
-
-template <class _RandomAccessIterator>
-inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20
-void sort(_RandomAccessIterator __first, _RandomAccessIterator __last) {
-  std::sort(__first, __last, __less<>());
-}
-
 _LIBCPP_END_NAMESPACE_STD
 
 #endif // _LIBCPP___ALGORITHM_SORT_H
diff --git a/include/algorithm b/include/algorithm
index 69ba953..97408ac 100644
--- a/include/algorithm
+++ b/include/algorithm
@@ -1949,6 +1949,7 @@ template <class BidirectionalIterator, class Compare>
 #include <__algorithm/shuffle.h>
 #include <__algorithm/sift_down.h>
 #include <__algorithm/sort.h>
+#include <__algorithm/sort_old.h>
 #include <__algorithm/sort_heap.h>
 #include <__algorithm/stable_partition.h>
 #include <__algorithm/stable_sort.h>
