diff --git a/contrib/ydb/core/cms/console/console__drop_yaml_config.cpp b/contrib/ydb/core/cms/console/console__drop_yaml_config.cpp
index 15accb2722..7fbaf09544 100644
--- a/contrib/ydb/core/cms/console/console__drop_yaml_config.cpp
+++ b/contrib/ydb/core/cms/console/console__drop_yaml_config.cpp
@@ -32,11 +32,15 @@ public:
             }
 
             if (Self->ClusterName != cluster) {
-                ythrow yexception() << "ClusterName mismatch";
+                ythrow yexception() << "ClusterName mismatch"
+                    << " expected " << Self->ClusterName
+                    << " but got " << cluster;
             }
 
             if (Version != Self->YamlVersion) {
-                ythrow yexception() << "Version mismatch";
+                ythrow yexception() << "Version mismatch"
+                    << " expected " << Self->YamlVersion
+                    << " but got " << Version;
             }
         } catch (const yexception& ex) {
             Error = true;
diff --git a/contrib/ydb/core/cms/console/console_configs_manager.cpp b/contrib/ydb/core/cms/console/console_configs_manager.cpp
index d9ba8fb42c..cac5a3cea1 100644
--- a/contrib/ydb/core/cms/console/console_configs_manager.cpp
+++ b/contrib/ydb/core/cms/console/console_configs_manager.cpp
@@ -79,11 +79,15 @@ void TConfigsManager::ValidateMainConfig(TUpdateConfigOpContext& opCtx) {
             auto resolved = NYamlConfig::ResolveAll(tree);
 
             if (ClusterName != opCtx.Cluster) {
-                ythrow yexception() << "ClusterName mismatch";
+                ythrow yexception() << "ClusterName mismatch"
+                    << " expected " << ClusterName
+                    << " but got " << opCtx.Cluster;
             }
 
             if (opCtx.Version != YamlVersion) {
-                ythrow yexception() << "Version mismatch";
+                ythrow yexception() << "Version mismatch"
+                    << " expected " << YamlVersion
+                    << " but got " << opCtx.Version;
             }
 
             TSimpleSharedPtr<NYamlConfig::TBasicUnknownFieldsCollector> unknownFieldsCollector = new NYamlConfig::TBasicUnknownFieldsCollector;
@@ -1070,11 +1074,15 @@ void TConfigsManager::Handle(TEvConsole::TEvAddVolatileConfigRequest::TPtr &ev,
             }
 
             if (ClusterName != clusterName) {
-                ythrow yexception() << "ClusterName mismatch";
+                ythrow yexception() << "ClusterName mismatch"
+                    << " expected " << ClusterName
+                    << " but got " << clusterName;
             }
 
             if (YamlVersion != version) {
-                ythrow yexception() << "Version mismatch";
+                ythrow yexception() << "Version mismatch"
+                    << " expected " << YamlVersion
+                    << " but got " << version;
             }
 
             VolatileYamlConfigs.try_emplace(id, cfg);
@@ -1105,11 +1113,15 @@ void TConfigsManager::Handle(TEvConsole::TEvRemoveVolatileConfigRequest::TPtr &e
     try {
         if (!rec.force()) {
             if (ClusterName != rec.identity().cluster()) {
-                ythrow yexception() << "ClusterName mismatch";
+                ythrow yexception() << "ClusterName mismatch"
+                    << " expected " << ClusterName
+                    << " but got " << rec.identity().cluster();
             }
 
             if (YamlVersion != rec.identity().version()) {
-                ythrow yexception() << "Version mismatch";
+                ythrow yexception() << "Version mismatch"
+                    << " expected " << YamlVersion
+                    << " but got " << rec.identity().version();
             }
         }
 
