diff --git a/contrib/ydb/library/yaml_config/public/yaml_config.cpp b/contrib/ydb/library/yaml_config/public/yaml_config.cpp
index 0af80634e5..2bdaf1b30f 100644
--- a/contrib/ydb/library/yaml_config/public/yaml_config.cpp
+++ b/contrib/ydb/library/yaml_config/public/yaml_config.cpp
@@ -881,6 +881,20 @@ std::variant<TMainMetadata, TDatabaseMetadata, TError> GetGenericMetadata(const
     }
 }
 
+TString UpgradeMainConfigVersion(const TString& config) {
+    auto metadata = GetMainMetadata(config);
+    Y_ENSURE(metadata.Version);
+    *metadata.Version = *metadata.Version + 1;
+    return ReplaceMetadata(config, metadata);
+}
+
+TString UpgradeStorageConfigVersion(const TString& config) {
+    auto metadata = GetStorageMetadata(config);
+    Y_ENSURE(metadata.Version);
+    *metadata.Version = *metadata.Version + 1;
+    return ReplaceMetadata(config, metadata);
+}
+
 } // namespace NKikimr::NYamlConfig
 
 template <>
diff --git a/contrib/ydb/library/yaml_config/public/yaml_config.h b/contrib/ydb/library/yaml_config/public/yaml_config.h
index 3a1e8111cf..3ff6f7a05f 100644
--- a/contrib/ydb/library/yaml_config/public/yaml_config.h
+++ b/contrib/ydb/library/yaml_config/public/yaml_config.h
@@ -245,6 +245,16 @@ TVolatileMetadata GetVolatileMetadata(const TString& config);
  */
 TString ReplaceMetadata(const TString& config, const TMainMetadata& metadata);
 
+/**
+ * Takes valid MainConfig and increases version exactly by one
+ */
+TString UpgradeMainConfigVersion(const TString& config);
+
+/**
+ * Takes valid MainConfig and increases version exactly by one
+ */
+TString UpgradeStorageConfigVersion(const TString& config);
+
 /**
  * Replaces metadata in database config
  */
diff --git a/contrib/ydb/public/lib/ydb_cli/commands/ydb_admin.cpp b/contrib/ydb/public/lib/ydb_cli/commands/ydb_admin.cpp
index d77416386b..18be0fb4ec 100644
--- a/contrib/ydb/public/lib/ydb_cli/commands/ydb_admin.cpp
+++ b/contrib/ydb/public/lib/ydb_cli/commands/ydb_admin.cpp
@@ -31,14 +31,14 @@ public:
     TCommandDatabase()
         : TClientCommandTree("database", {}, "Database-wide administration")
     {
-        AddCommand(std::make_unique<NDynamicConfig::TCommandConfig>());
+        AddCommand(std::make_unique<NDynamicConfig::TCommandConfig>(false));
         AddCommand(std::make_unique<TCommandDatabaseDump>());
         AddCommand(std::make_unique<TCommandDatabaseRestore>());
     }
 };
 
-TCommandDatabaseDump::TCommandDatabaseDump() 
-    : TYdbReadOnlyCommand("dump", {}, "Dump database into local directory") 
+TCommandDatabaseDump::TCommandDatabaseDump()
+    : TYdbReadOnlyCommand("dump", {}, "Dump database into local directory")
 {}
 
 void TCommandDatabaseDump::Config(TConfig& config) {
@@ -66,8 +66,8 @@ int TCommandDatabaseDump::Run(TConfig& config) {
     return EXIT_SUCCESS;
 }
 
-TCommandDatabaseRestore::TCommandDatabaseRestore() 
-    : TYdbCommand("restore", {}, "Restore database from local dump") 
+TCommandDatabaseRestore::TCommandDatabaseRestore()
+    : TYdbCommand("restore", {}, "Restore database from local dump")
 {}
 
 void TCommandDatabaseRestore::Config(TConfig& config) {
@@ -109,6 +109,7 @@ TCommandAdmin::TCommandAdmin()
     UseOnlyExplicitProfile();
     // keep old commands "safe", to keep old behavior
     AddHiddenCommand(std::make_unique<NDynamicConfig::TCommandConfig>(
+                         true,
                          NDynamicConfig::TCommandFlagsOverrides{.Dangerous = false, .OnlyExplicitProfile = false},
                          false));
     AddHiddenCommand(std::make_unique<NDynamicConfig::TCommandVolatileConfig>());
diff --git a/contrib/ydb/public/lib/ydb_cli/commands/ydb_cluster.cpp b/contrib/ydb/public/lib/ydb_cli/commands/ydb_cluster.cpp
index 8de1d17af6..f2b9a5a8a3 100644
--- a/contrib/ydb/public/lib/ydb_cli/commands/ydb_cluster.cpp
+++ b/contrib/ydb/public/lib/ydb_cli/commands/ydb_cluster.cpp
@@ -17,7 +17,7 @@ TCommandCluster::TCommandCluster()
     : TClientCommandTree("cluster", {}, "Cluster-wide administration")
 {
     AddCommand(std::make_unique<TCommandClusterBootstrap>());
-    AddCommand(std::make_unique<NDynamicConfig::TCommandConfig>(true));
+    AddCommand(std::make_unique<NDynamicConfig::TCommandConfig>(false, true));
     AddCommand(std::make_unique<TCommandClusterDump>());
     AddCommand(std::make_unique<TCommandClusterRestore>());
 }
diff --git a/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.cpp b/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.cpp
index 3a5c38cfce..2d01e02471 100644
--- a/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.cpp
+++ b/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.cpp
@@ -39,19 +39,22 @@ TString WrapStaticConfig(const TString& yaml) {
 }
 
 TCommandConfig::TCommandConfig(
+        bool useLegacyApi,
         TCommandFlagsOverrides commandFlagsOverrides,
         bool allowEmptyDatabase)
     : TClientCommandTree("config", {}, "Dynamic config")
     , CommandFlagsOverrides(commandFlagsOverrides)
 {
-    AddCommand(std::make_unique<TCommandConfigFetch>(allowEmptyDatabase));
-    AddCommand(std::make_unique<TCommandConfigReplace>(allowEmptyDatabase));
+    AddCommand(std::make_unique<TCommandConfigFetch>(useLegacyApi, allowEmptyDatabase));
+    AddCommand(std::make_unique<TCommandConfigReplace>(useLegacyApi, allowEmptyDatabase));
     AddCommand(std::make_unique<TCommandConfigResolve>());
     AddCommand(std::make_unique<TCommandGenerateDynamicConfig>(allowEmptyDatabase));
 }
 
-TCommandConfig::TCommandConfig(bool allowEmptyDatabase)
-    : TCommandConfig(TCommandFlagsOverrides{}, allowEmptyDatabase)
+TCommandConfig::TCommandConfig(
+    bool useLegacyApi,
+    bool allowEmptyDatabase)
+    : TCommandConfig(useLegacyApi, TCommandFlagsOverrides{}, allowEmptyDatabase)
 {}
 
 void TCommandConfig::PropagateFlags(const TCommandFlags& flags) {
@@ -70,8 +73,11 @@ void TCommandConfig::PropagateFlags(const TCommandFlags& flags) {
     }
 }
 
-TCommandConfigFetch::TCommandConfigFetch(bool allowEmptyDatabase)
+TCommandConfigFetch::TCommandConfigFetch(
+        bool useLegacyApi,
+        bool allowEmptyDatabase)
     : TYdbReadOnlyCommand("fetch", {"get", "dump"}, "Fetch main dynamic-config")
+    , UseLegacyApi(useLegacyApi)
     , AllowEmptyDatabase(allowEmptyDatabase)
 {
 }
@@ -106,11 +112,16 @@ int TCommandConfigFetch::Run(TConfig& config) {
     auto driver = std::make_unique<NYdb::TDriver>(CreateDriver(config));
     auto client = NYdb::NConfig::TConfigClient(*driver);
 
-    NYdb::NConfig::TFetchAllConfigsSettings settings;
-    auto result = client.FetchAllConfigs(settings).GetValueSync();
+    NYdb::NConfig::TFetchConfigResult result(TStatus(EStatus::CLIENT_CALL_UNIMPLEMENTED, {}), {});
+
+    if (!UseLegacyApi) {
+        NYdb::NConfig::TFetchAllConfigsSettings settings;
+        result = client.FetchAllConfigs(settings).GetValueSync();
+    }
 
     // if the new Config API is not supported, fallback to the old DynamicConfig API
     if (result.GetStatus() == EStatus::CLIENT_CALL_UNIMPLEMENTED || result.GetStatus() == EStatus::UNSUPPORTED) {
+        Cerr << "Warning: Fallback to DynamicConfig API" << Endl;
         auto client = NYdb::NDynamicConfig::TDynamicConfigClient(*driver);
         auto result = client.GetConfig().GetValueSync();
         NStatusHelpers::ThrowOnErrorOrPrintIssues(result);
@@ -160,7 +171,17 @@ int TCommandConfigFetch::Run(TConfig& config) {
         }, entry.Identity);
     }
 
+    // TODO: rewrite it in proper way, without stdout and stderr confusion
     if (!clusterConfig.empty()) {
+        try {
+            clusterConfig = NYamlConfig::UpgradeMainConfigVersion(clusterConfig);
+        } catch(...) {
+            // it is better to return at least something
+            // because for user it is the only way to get config
+            // user will be unable to reupload this config without manual interaction
+            // and will get attention that something went horribly wrong
+            Cerr << "Unable to bump main config version, returning as-is" << Endl;
+        }
         if (!storageConfig.empty() || DedicatedStorageSection) {
             Cerr << "cluster config: " << Endl;
         }
@@ -168,6 +189,15 @@ int TCommandConfigFetch::Run(TConfig& config) {
     }
 
     if (!storageConfig.empty()) {
+        try {
+            storageConfig = NYamlConfig::UpgradeStorageConfigVersion(storageConfig);
+        } catch(...) {
+            // it is better to return at least something
+            // because for user it is the only way to get config
+            // user will be unable to reupload this config without manual interaction
+            // and will get attention that something went horribly wrong
+            Cerr << "Unable to bump storage config version, returning as-is" << Endl;
+        }
         if (!clusterConfig.empty() || DedicatedClusterSection) {
             Cerr << "storage config:" << Endl;
         }
@@ -182,8 +212,11 @@ int TCommandConfigFetch::Run(TConfig& config) {
     return EXIT_SUCCESS;
 }
 
-TCommandConfigReplace::TCommandConfigReplace(bool allowEmptyDatabase)
+TCommandConfigReplace::TCommandConfigReplace(
+        bool useLegacyApi,
+        bool allowEmptyDatabase)
     : TYdbCommand("replace", {}, "Replace dynamic config")
+    , UseLegacyApi(useLegacyApi)
     , IgnoreCheck(false)
     , AllowEmptyDatabase(allowEmptyDatabase)
 {
@@ -242,7 +275,11 @@ int TCommandConfigReplace::Run(TConfig& config) {
         settings.AllowUnknownFields();
     }
 
-    auto status = client.ReplaceConfig(DynamicConfig, settings).GetValueSync();
+    auto status = TStatus(EStatus::CLIENT_CALL_UNIMPLEMENTED, {});
+
+    if (!UseLegacyApi) {
+        status = client.ReplaceConfig(DynamicConfig, settings).GetValueSync();
+    }
 
     if (status.GetStatus() == EStatus::CLIENT_CALL_UNIMPLEMENTED || status.GetStatus() == EStatus::UNSUPPORTED) {
         Cerr << "Warning: Fallback to DynamicConfig API" << Endl;
diff --git a/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.h b/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.h
index b21a8168ee..fbdff178d7 100644
--- a/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.h
+++ b/contrib/ydb/public/lib/ydb_cli/commands/ydb_dynamic_config.h
@@ -15,10 +15,13 @@ struct TCommandFlagsOverrides {
 class TCommandConfig : public TClientCommandTree {
 public:
     TCommandConfig(
+        bool useLegacyApi,
         TCommandFlagsOverrides commandFlagsOverrides = {},
         bool allowEmptyDatabase = false);
 
-    TCommandConfig(bool allowEmptyDatabase);
+    TCommandConfig(
+        bool useLegacyApi,
+        bool allowEmptyDatabase);
 
     void PropagateFlags(const TCommandFlags& flags) override;
 private:
@@ -27,12 +30,15 @@ private:
 
 class TCommandConfigReplace : public TYdbCommand {
 public:
-    TCommandConfigReplace(bool allowEmptyDatabase);
+    TCommandConfigReplace(
+        bool useLegacyApi,
+        bool allowEmptyDatabase);
     void Config(TConfig& config) override;
     void Parse(TConfig& config) override;
     int Run(TConfig& config) override;
 
 private:
+    bool UseLegacyApi = false;
     bool IgnoreCheck = false;
     bool Force = false;
     bool DryRun = false;
@@ -44,12 +50,15 @@ private:
 
 class TCommandConfigFetch : public TYdbReadOnlyCommand {
 public:
-    TCommandConfigFetch(bool allowEmptyDatabase);
+    TCommandConfigFetch(
+        bool useLegacyApi,
+        bool allowEmptyDatabase);
     void Config(TConfig&) override;
     void Parse(TConfig&) override;
     int Run(TConfig& config) override;
 
 private:
+    bool UseLegacyApi = false;
     bool StripMetadata = false;
     TString OutDir;
     bool AllowEmptyDatabase = false;
