diff --git a/contrib/ydb/library/yaml_config/public/yaml_config.cpp b/contrib/ydb/library/yaml_config/public/yaml_config.cpp
index 0af80634e5..a713851fe4 100644
--- a/contrib/ydb/library/yaml_config/public/yaml_config.cpp
+++ b/contrib/ydb/library/yaml_config/public/yaml_config.cpp
@@ -636,7 +636,7 @@ description: Implicit DatabaseConfig node
 selector: {}
 )"));
     auto node = databaseConfigRoot.Map()["config"].Copy(config);
-    selectors.at(0).Map().Append(config.Buildf("config"), node);
+    selectors.at(selectors.size() - 1).Map().Append(config.Buildf("config"), node);
 }
 
 ui64 GetVersion(const TString& config) {
diff --git a/contrib/ydb/library/yaml_config/yaml_config_ut.cpp b/contrib/ydb/library/yaml_config/yaml_config_ut.cpp
index 42831cecbb..ec9a143f62 100644
--- a/contrib/ydb/library/yaml_config/yaml_config_ut.cpp
+++ b/contrib/ydb/library/yaml_config/yaml_config_ut.cpp
@@ -1473,6 +1473,147 @@ Y_UNIT_TEST_SUITE(YamlConfig) {
         stream << cfg;
     }
 
+    void AppendDatabaseConfigBase(
+        const TString mainConfig,
+        const TString databaseConfig)
+    {
+        auto treeOriginal = NFyaml::TDocument::Parse(mainConfig);
+        auto tree = NFyaml::TDocument::Parse(mainConfig);
+        auto dbTree = NFyaml::TDocument::Parse(databaseConfig);
+
+        const size_t selectorsCountBefore =
+            tree.Root().Map().Has("selector_config")
+                ? tree.Root().Map().at("selector_config").Sequence().size()
+                : 0;
+
+        // Add empty selector_config nodes to avoid processing missing YAML
+        // nodes
+        if (!tree.Root().Map().Has("selector_config")) {
+            tree.Root().Map().Append(
+                tree.Buildf("selector_config"),
+                tree.Buildf("[]"));
+        }
+        if (!treeOriginal.Root().Map().Has("selector_config")) {
+            treeOriginal.Root().Map().Append(
+                treeOriginal.Buildf("selector_config"),
+                treeOriginal.Buildf("[]"));
+        }
+
+        UNIT_ASSERT_NO_EXCEPTION(
+            NKikimr::NYamlConfig::AppendDatabaseConfig(tree, dbTree));
+
+        auto selectors = tree.Root().Map().at("selector_config").Sequence();
+        UNIT_ASSERT_VALUES_EQUAL(selectors.size(), selectorsCountBefore + 1);
+
+        auto lastSelector = selectors.at(selectors.size() - 1).Map();
+        UNIT_ASSERT(lastSelector.Has("config"));
+
+        // Check that main config selectors have not changed
+        auto selectorsOriginal =
+            treeOriginal.Root().Map().at("selector_config").Sequence();
+
+        for (size_t i = 0; i < selectorsCountBefore; i++) {
+            UNIT_ASSERT_VALUES_EQUAL(
+                selectors.at(i).Map().size(),
+                selectorsOriginal.at(i).Map().size());
+
+            // Сheck all map elements (description, selector, config)
+            for (auto field = selectors.at(i).Map().begin(),
+                      fieldOrig = selectorsOriginal.at(i).Map().begin();
+                 field != selectors.at(i).Map().end();
+                 field++, fieldOrig++)
+            {
+                TStringStream fieldStr;
+                fieldStr << field->Key() << ": " << field->Value();
+
+                TStringStream fieldOriginalStr;
+                fieldOriginalStr << fieldOrig->Key() << ": "
+                                 << fieldOrig->Value();
+
+                UNIT_ASSERT_VALUES_EQUAL(
+                    fieldStr.Str(),
+                    fieldOriginalStr.Str());
+            }
+        }
+
+        // Check that the db config selector is appended to the end of the main
+        // config selectors
+        TStringStream actualLastConfig;
+        actualLastConfig << lastSelector.at("config");
+
+        TStringStream expectedConfig;
+        expectedConfig << dbTree.Root().Map().at("config");
+
+        UNIT_ASSERT_VALUES_EQUAL(actualLastConfig.Str(), expectedConfig.Str());
+    }
+
+    Y_UNIT_TEST(AppendDatabaseConfigAsTheOnlySelector)
+    {
+        const TString mainConfig = R"(
+---
+metadata:
+  kind: MainConfig
+  cluster: ""
+  version: 0
+config:
+  feature_flags:
+    some_param_1: true
+)";
+
+        const TString databaseConfig = R"(
+---
+metadata:
+  kind: DatabaseConfig
+  database: "/dc/tenant"
+  version: 0
+config:
+  feature_flags: !inherit
+    some_param_1: false
+)";
+
+        AppendDatabaseConfigBase(mainConfig, databaseConfig);
+    }
+
+    Y_UNIT_TEST(AppendDatabaseConfigAfterExistingSelectors)
+    {
+        const TString mainConfig = R"(
+---
+metadata:
+  kind: MainConfig
+  cluster: ""
+  version: 0
+config:
+  feature_flags:
+    some_param_1: true
+selector_config:
+- description: pre-existing selector 1 with config
+  selector:
+    tenant: dynamic
+  config:
+    feature_flags: !inherit
+      some_param_2: true
+- description: pre-existing selector 2 with config
+  selector:
+    tenant: dynamic
+  config:
+    feature_flags: !inherit
+      some_param_3: true
+)";
+
+        const TString databaseConfig = R"(
+---
+metadata:
+  kind: DatabaseConfig
+  database: "/dc/tenant"
+  version: 0
+config:
+  feature_flags: !inherit
+    some_param_1: false
+)";
+
+        AppendDatabaseConfigBase(mainConfig, databaseConfig);
+    }
+
     Y_UNIT_TEST(GetMetadata) {
         {
             TString str = R"(
