--- contrib/python/aiohttp/aiohttp/client.py	(index)
+++ contrib/python/aiohttp/aiohttp/client.py	(working tree)
@@ -566,6 +566,8 @@ class ClientSession:
                     except ClientError:
                         raise
                     except OSError as exc:
+                        if exc.errno is None and isinstance(exc, asyncio.TimeoutError):
+                            raise
                         raise ClientOSError(*exc.args) from exc
 
                     self._cookie_jar.update_cookies(resp.cookies, resp.url)
