commit 72caba30262ecd4c40539313670bc8d56971e58e (pydebug, arcadia/users/shadchin/pydebug)
author: shadchin
date: 2023-03-30T07:20:46+03:00

    XXX protobuf

--- contrib/python/protobuf/py3/google/protobuf/pyext/descriptor.cc	(e74823f2384f6b0811c1c9982c1c644ed526a4d0)
+++ contrib/python/protobuf/py3/google/protobuf/pyext/descriptor.cc	(72caba30262ecd4c40539313670bc8d56971e58e)
@@ -449,6 +449,9 @@ PyObject* NewInternedDescriptor(PyTypeObject* type,
 
 static void Dealloc(PyObject* pself) {
   PyBaseDescriptor* self = reinterpret_cast<PyBaseDescriptor*>(pself);
+  if (PyObject_GC_IsTracked(pself)) {
+    PyObject_GC_UnTrack(pself);
+  }
   // Remove from interned dictionary
   interned_descriptors->erase(self->descriptor);
   Py_CLEAR(self->pool);
--- contrib/python/protobuf/py3/google/protobuf/pyext/descriptor_pool.cc	(e74823f2384f6b0811c1c9982c1c644ed526a4d0)
+++ contrib/python/protobuf/py3/google/protobuf/pyext/descriptor_pool.cc	(72caba30262ecd4c40539313670bc8d56971e58e)
@@ -199,6 +199,9 @@ static PyObject* New(PyTypeObject* type,
 
 static void Dealloc(PyObject* pself) {
   PyDescriptorPool* self = reinterpret_cast<PyDescriptorPool*>(pself);
+  if (PyObject_GC_IsTracked(pself)) {
+    PyObject_GC_UnTrack(pself);
+  }
   descriptor_pool_map->erase(self->pool);
   Py_CLEAR(self->py_message_factory);
   for (std::unordered_map<const void*, PyObject*>::iterator it =
--- contrib/python/protobuf/py3/google/protobuf/pyext/map_container.cc	(e74823f2384f6b0811c1c9982c1c644ed526a4d0)
+++ contrib/python/protobuf/py3/google/protobuf/pyext/map_container.cc	(72caba30262ecd4c40539313670bc8d56971e58e)
@@ -521,6 +521,9 @@ PyObject* MapReflectionFriend::ScalarMapToStr(PyObject* _self) {
 
 static void ScalarMapDealloc(PyObject* _self) {
   MapContainer* self = GetMap(_self);
+  if (PyObject_GC_IsTracked(_self)) {
+    PyObject_GC_UnTrack(_self);
+  }
   self->RemoveFromParentCache();
   PyTypeObject *type = Py_TYPE(_self);
   type->tp_free(_self);
@@ -787,6 +790,9 @@ PyObject* MessageMapGet(PyObject* self, PyObject* args, PyObject* kwargs) {
 
 static void MessageMapDealloc(PyObject* _self) {
   MessageMapContainer* self = GetMessageMap(_self);
+  if (PyObject_GC_IsTracked(_self)) {
+    PyObject_GC_UnTrack(_self);
+  }
   self->RemoveFromParentCache();
   Py_DECREF(self->message_class);
   PyTypeObject *type = Py_TYPE(_self);
--- contrib/python/protobuf/py3/google/protobuf/pyext/message_factory.cc	(e74823f2384f6b0811c1c9982c1c644ed526a4d0)
+++ contrib/python/protobuf/py3/google/protobuf/pyext/message_factory.cc	(72caba30262ecd4c40539313670bc8d56971e58e)
@@ -106,6 +106,9 @@ PyObject* New(PyTypeObject* type, PyObject* args, PyObject* kwargs) {
 static void Dealloc(PyObject* pself) {
   PyMessageFactory* self = reinterpret_cast<PyMessageFactory*>(pself);
 
+  if (PyObject_GC_IsTracked(pself)) {
+    PyObject_GC_UnTrack(pself);
+  }
   typedef PyMessageFactory::ClassesByMessageMap::iterator iterator;
   for (iterator it = self->classes_by_descriptor->begin();
        it != self->classes_by_descriptor->end(); ++it) {
