diff --git a/ares_setup.h b/ares_setup.h
index 4df7961..b9a1fd5 100644
--- a/src/lib/ares_setup.h
+++ b/src/lib/ares_setup.h
@@ -28,1 +28,1 @@
-#ifdef HAVE_CONFIG_H
+#if defined(HAVE_CONFIG_H) && !defined(_MSC_VER)
diff --git a/setup_once.h b/setup_once.h
index a8cfe6b..d4b0a95 100644
--- a/src/lib/setup_once.h
+++ b/src/lib/setup_once.h
@@ -122,1 +122,1 @@ struct timeval {
-#ifdef HAVE_MSG_NOSIGNAL
+#if defined(HAVE_MSG_NOSIGNAL) && defined(MSG_NOSIGNAL)
