commit 1efd2cc9274f1f5a222e759093d27c8d38825028 (HEAD -> yamaker-bison, arcadia/users/thegeorg/yamaker-bison)
author: thegeorg
date: 2024-06-18T12:17:40+03:00

    Put contrib/tools/bison under yamaker

--- contrib/tools/bison/src/getargs.c	(ef06a7dccff68bd15f669606131c541d158d4b1a)
+++ contrib/tools/bison/src/getargs.c	(1efd2cc9274f1f5a222e759093d27c8d38825028)
@@ -365,15 +363,6 @@ FEATURE is a list of comma separated words that can include:\n\
       /* Don't output this redundant message for English locales.
          Note we still output for 'C' so that it gets included in the
          man page.  */
-      const char *lc_messages = setlocale (LC_MESSAGES, NULL);
-      if (lc_messages && !STREQ (lc_messages, "en_"))
-        /* TRANSLATORS: Replace LANG_CODE in this URL with your language
-           code <http://translationproject.org/team/LANG_CODE.html> to
-           form one of the URLs at http://translationproject.org/team/.
-           Otherwise, replace the entire URL with your translation team's
-           email address.  */
-        fputs (_("Report translation bugs to "
-                 "<http://translationproject.org/team/>.\n"), stdout);
       fputs (_("For complete documentation, run: info bison.\n"), stdout);
     }
 
--- contrib/tools/bison/src/main.c	(ef06a7dccff68bd15f669606131c541d158d4b1a)
+++ contrib/tools/bison/src/main.c	(1efd2cc9274f1f5a222e759093d27c8d38825028)
@@ -60,9 +60,6 @@ main (int argc, char *argv[])
 {
   set_program_name (argv[0]);
   setlocale (LC_ALL, "");
-  (void) bindtextdomain (PACKAGE, LOCALEDIR);
-  (void) bindtextdomain ("bison-runtime", LOCALEDIR);
-  (void) textdomain (PACKAGE);
 
   {
     char const *cp = getenv ("LC_CTYPE");
